<?xml version="1.0" encoding="utf-8"?>

<launch>
  <param name="use_sim_time" value="false"/>
  <arg name="visualize" default="false"/>

  <arg name="play_rosbag" default="false"/>
  <arg name="rosbag_file"
       default="$(find cpt_object_detection)/resources/detection_pawn_sampseg.bag"/>

  <arg name="detection_pointcloud_topic"
       default="/detection_projector/object_pcl_3d"/>
  <arg name="object_off_path"
       default="$(find cpt_object_detection)/resources/handwheel_mockup_24cm_scaled.off"/>
  <arg name="object_type" default="valve"/>
  <arg name="object_frame_id" default="object_detection_mesh"/>
  <arg name="publish_static_transform" default="true"/>
  <arg name="reference_frame_id" default="tracking_cam_odom_frame"/>
  <arg name="initialization_duration_s" default="1.0"/>

  <!-- Filters -->
  <arg name="use_inlier_ratio_filter" default="false"/>
  <arg name="inlier_ratio_decay" default="0.95"/>
  <arg name="use_kalman_filter" default="false"/>
  <arg name="kf_p_translation" default="0"/>
  <arg name="kf_p_rotation" default="10"/>
  <arg name="kf_r_translation" default="0.1"/>
  <arg name="kf_r_rotation" default="1"/>

  <arg name="use_3d_features" default="false"/>
  <arg name="num_points_object_pointcloud" default="5000"/>
  <arg name="downsampling" default="0.00"/>
  <arg name="keypoint_type" default="uniform"/> <!-- ISS or Harris or uniform or all -->
  <arg name="descriptor_type" default="unit"/> <!-- FPFH or SHOT or 3DSmoothNet or unit -->
  <arg name="matching_method" default="Teaser"/> <!-- geometric_consistency or FGR or Teaser -->
  <arg name="correspondence_threshold" default="0.1"/>

  <arg name="use_icp_on_pointcloud" default="true"/>
  <arg name="refine_using_icp" default="true"/>
  <arg name="icp_config_file"
       default="$(find cpt_object_detection)/config/icp_config.yaml"/>

  <node name="object_detection_matcher"
        pkg="cpt_object_detection" type="cpt_object_detection_node"
        args="--colorlogtostderr" output="screen">
    <param name="pointcloud_topic" value="$(arg detection_pointcloud_topic)"/>
    <param name="off_model" value="$(arg object_off_path)"/>
    <param name="visualize_object_on_startup" value="true"/>
    <param name="object_frame_id" value="$(arg object_frame_id)"/>
    <param name="publish_static_transform" value="$(arg publish_static_transform)"/>
    <param name="reference_frame_id" value="$(arg reference_frame_id)"/>
    <param name="initialization_duration_s" value="$(arg initialization_duration_s)"/>

    <param name="use_inlier_ratio_filter" value="$(arg use_inlier_ratio_filter)"/>
    <param name="inlier_ratio_decay" value="$(arg use_inlier_ratio_filter)"/>
    <param name="use_kalman_filter" value="$(arg use_kalman_filter)"/>
    <param name="kf_p_translation" value="$(arg kf_p_translation)"/>
    <param name="kf_p_rotation" value="$(arg kf_p_rotation)"/>
    <param name="kf_r_translation" value="$(arg kf_r_translation)"/>
    <param name="kf_r_rotation" value="$(arg kf_r_rotation)"/>

    <param name="use_3d_features" value="$(arg use_3d_features)"/>
    <param name="num_points_object_pointcloud"
           value="$(arg num_points_object_pointcloud)"/>
    <param name="downsampling" value="$(arg downsampling)"/>
    <param name="keypoint_type" value="$(arg keypoint_type)"/>
    <param name="descriptor_type" value="$(arg descriptor_type)"/>
    <param name="matching_method" value="$(arg matching_method)"/>
    <param name="correspondence_threshold" value="$(arg correspondence_threshold)"/>

    <param name="refine_using_icp" value="$(arg refine_using_icp)"/>
    <param name="use_icp_on_pointcloud" value="$(arg use_icp_on_pointcloud)"/>
    <param name="icp_config_file" value="$(arg icp_config_file)"/>
  </node>

  <node if="$(arg play_rosbag)" name="rosbag_player" pkg="rosbag" type="play"
        args="--clock -q $(arg rosbag_file)"/>

  <node if="$(arg visualize)" name="rviz" pkg="rviz" type="rviz"
        args="-d $(find cpt_object_detection)/config/object_detection_matching_3d_features.rviz"/>
</launch>
